dotnet/src/SemanticKernel/AI/Embeddings/VectorOperations/DotProductOperation.cs

```diff
@@ -75,237 +76,76 @@ private static unsafe double DotProductImplementation(ReadOnlySpan<double> x, Re
             throw new ArgumentException("Array lengths must be equal");
         }
 
-        if (x.Length % 4 == 0)
+        fixed (double* pxBuffer = x, pyBuffer = y)
         {
-            return DotProduct_Len4(x, y);
-        }
+            double* px = pxBuffer, py = pyBuffer;
+            double* pxEnd = px + x.Length;
 
-        if (x.Length % 2 == 0)
-        {
-            return DotProduct_Len2(x, y);
-        }
+            double dotSum = 0;
 
-        // Vanilla Dot Product
-        fixed (double* pxBuffer = x)
-        {
-            fixed (double* pyBuffer = y)
+            if (Vector.IsHardwareAccelerated &&
+                x.Length >= Vector<double>.Count)
             {
-                double dotSum = 0;
-                double* px = pxBuffer;
-                double* pxMax = px + x.Length;
-                double* py = pyBuffer;
-                while (px < pxMax)
+                double* pxOneVectorFromEnd = pxEnd - Vector<double>.Count;
+                do
                 {
-                    // Dot product
-                    dotSum += *px * *py;
-                    ++px;
-                    ++py;
-                }
-
-                return dotSum;
-            }
-        }
-    }
-
-    private static unsafe double DotProductImplementation(ReadOnlySpan<float> x, ReadOnlySpan<float> y)
-    {
-        if (x.Length != y.Length)
-        {
-            throw new ArgumentException("Array lengths must be equal");
-        }
-
-        if (x.Length % 4 == 0)
-        {
-            return DotProduct_Len4(x, y);
-        }
-
-        if (x.Length % 2 == 0)
-        {
-            // Twice as fast
-            return DotProduct_Len2(x, y);
-        }
+                    dotSum += Vector.Dot(*(Vector<double>*)px, *(Vector<double>*)py); // Dot product
```

Comment:
Since the input is ReadOnlySpan<float>, this looks fishy:
*(Vector<double>*)px, 
Should it be *(Vector<float>) ?